### PR TITLE
MODCAMUNDA-16: Use Kafka

### PR DESCRIPTION
**Pull request by wwelling**
Requires https://github.com/folio-org/spring-module-core/pull/67

[MODCAMUNDA-16](https://folio-org.atlassian.net/browse/MODCAMUNDA-16)

- add history API to module descriptor
- fix workflow engine rest path to camunda

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -101,7 +104,8 @@ spring:
     mode: TEXT
     suffix: .sql
 
-  jersey.application-path: rest
+  # this must be `camunda` to match module descriptor path patterns
+  jersey.application-path: camunda
```
Comment (wwelling) [inline src/main/resources/application.yaml:108]: This must match module descriptor and probably should be a template variable. For now, hardcoding it.

Comment (wwelling) [reply src/main/resources/application.yaml:108]: If and when changed, mod-workflow property `okapi.workflow-engine.path` must be updated accordingly. 

https://github.com/folio-org/mod-workflow/pull/116/files#diff-5f1399fae635ac57e696d503eb2135f7575b3a59a486e21613d3343f8f2a3243R50

**Review by Dbreck-TAMU — APPROVED** [review]

**Review by jeremythuff — APPROVED** [review]

File: descriptors/ModuleDescriptor-template.json [hunk 2]
```diff
@@ -2,6 +2,22 @@
   "id": "@project.artifactId@-@project.version@",
   "name": "Camnuda BPM Module",
   "provides": [
+    {
+      "id": "camunda-history",
```
Comment (kaladay) [inline descriptors/ModuleDescriptor-template.json:6]: Does Kafka require this or is this a new feature being added?

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -365,11 +401,16 @@
...
       { "name": "DB_CHARSET", "value": "UTF-8" },
       { "name": "DB_MAXPOOLSIZE", "value": "16" },
-      { "name": "EVENT_QUEUE_NAME", "value": "event.queue", "description": "The ActiveMQ event queue name." },
+      { "name": "KAFKA_HOST", "value": "kafka" },
+      { "name": "KAFKA_PORT", "value": "9092" },
+      { "name": "KAFKA_SECURITY_PROTOCOL", "value": "PLAINTEXT" },
+      { "name": "KAFKA_SSL_KEYSTORE_LOCATION", "value": "" },
+      { "name": "KAFKA_SSL_KEYSTORE_PASSWORD", "value": "" },
+      { "name": "KAFKA_SSL_TRUSTSTORE_LOCATION", "value": "" },
+      { "name": "KAFKA_SSL_TRUSTSTORE_PASSWORD", "value": "" },
```
Comment (kaladay) [inline descriptors/ModuleDescriptor-template.json:410]: Each of these must have a description field.
Please add a (non-empty) description field.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java [hunk 4]
```diff
@@ -28,13 +28,13 @@ public WorkflowController(CamundaApiService camundaApiService) {
   @PostMapping(value = {"/activate", "/activate/"}, produces = { MediaType.APPLICATION_JSON_VALUE })
   public Workflow activateWorkflow(@RequestBody Workflow workflow, @TenantHeader String tenant)
       throws WorkflowAlreadyActiveException, ScriptTaskDeserializeCodeFailure {
-    log.debug("Activating Workflow: {}", workflow == null ? null : workflow.getId());
+    log.info("Activating Workflow: {}", workflow == null ? null : workflow.getId());
```
Comment (kaladay) [inline src/main/java/org/folio/rest/camunda/controller/WorkflowController.java:31]:
```suggestion
    log.debug("Activating Workflow: {}", workflow == null ? null : workflow.getId());
```

Should be `debug`.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java [hunk 5]
```diff
@@ -28,13 +28,13 @@ public WorkflowController(CamundaApiService camundaApiService) {
...
     return camundaApiService.deployWorkflow(workflow, tenant);
   }
 
   @PostMapping(value = {"/deactivate", "/deactivate/"}, produces = { MediaType.APPLICATION_JSON_VALUE })
   public Workflow deactivateWorkflow(@RequestBody Workflow workflow) {
-    log.debug("Deactivating Workflow: {}", workflow == null ? null : workflow.getId());
+    log.info("Deactivating Workflow: {}", workflow == null ? null : workflow.getId());
```
Comment (kaladay) [inline src/main/java/org/folio/rest/camunda/controller/WorkflowController.java:37]:
```suggestion
    log.debug("Deactivating Workflow: {}", workflow == null ? null : workflow.getId());
```

Should be `debug`.

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -134,13 +137,15 @@ camunda:
...
-    # see: https://docs.camunda.org/manual/7.20/user-guide/spring-boot-integration/configuration/#generic-properties 
+    # see: https://docs.camunda.org/manual/7.20/user-guide/spring-boot-integration/configuration/#generic-properties
     generic-properties:
       properties:
         telemetry-reporter-activate: false
         enable-script-compilation: true
         enable-script-engine-nashorn-compatibility: true
         enable-fetch-script-engine-from-process-application: false
+    metrics:
+      enabled: false
```
Comment (kaladay) [inline src/main/resources/application.yaml:148]: Wonderful!
That has been ever so annoying.

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -152,7 +157,13 @@ springdoc:
...
   writer-with-default-pretty-printer: true
 
-event.queue.name: event.queue
+application:
+  kafka:
+    listener:
+      events:
+        concurrency: 1
+        topic-pattern: (${ENV:folio}\.)?(.*\.)?workflow.events
+        group-id: ${ENV:folio}-mod-camunda-workflow-events-group
```
Comment (kaladay) [inline src/main/resources/application.yaml:166]: Is this saying the environment variable `folio`, all lower case?

This probably should be a normal environment variable and then be more descriptive, such as `FOLIO_NAMESPACE` (I randomly chose the word `NAMESPACE` in this example).

File: src/test/resources/application.yaml [hunk 8]
```diff
@@ -141,7 +144,13 @@ springdoc:
...
     path: /swagger-ui
 
-event.queue.name: event.queue
+application:
+  kafka:
+    listener:
+      events:
+        concurrency: 1
+        topic-pattern: (${ENV:folio}\.)?(.*\.)?workflow.events
+        group-id: ${ENV:folio}-mod-camunda-workflow-events-group
```
Comment (kaladay) [inline src/test/resources/application.yaml:153]: Same as noted above.

**Review by kaladay — CHANGES_REQUESTED** [review]

File: descriptors/ModuleDescriptor-template.json (hunk 2 above)
Comment (wwelling) [reply descriptors/ModuleDescriptor-template.json:6]: This is required for MIS Catalog reports and should be exposed.

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (wwelling) [reply descriptors/ModuleDescriptor-template.json:410]: Followed mod-remote-storage and will leave them as is until further review.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java (hunk 4 above)
Comment (wwelling) [reply src/main/java/org/folio/rest/camunda/controller/WorkflowController.java:31]: These are not debugging logs and should be aware of each workflow activation/deactivation as normal operating logs. The number of times these are actually logged is insignificant and can be info for useful activity logging.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java (hunk 5 above)
Comment (wwelling) [reply src/main/java/org/folio/rest/camunda/controller/WorkflowController.java:37]: Same as above.

File: src/main/resources/application.yaml (hunk 6 above)
Comment (wwelling) [reply src/main/resources/application.yaml:148]: Luckily it broke when deployed in folio/snapshot. Turning it off solves both.

File: src/main/resources/application.yaml (hunk 7 above)
Comment (wwelling) [reply src/main/resources/application.yaml:166]: This was following implementation details and properties conventions from mod-remote-storage.

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (kaladay) [reply descriptors/ModuleDescriptor-template.json:410]: I will not approve without the descriptions.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java (hunk 4 above)
Comment (kaladay) [reply src/main/java/org/folio/rest/camunda/controller/WorkflowController.java:31]: The previous code has them as `debug`.

I do not see how changing that is in the scope of this ticket.
I think excessive logging concerns should be a convesation for a different issue other than the Kafka.

File: src/main/resources/application.yaml (hunk 7 above)
Comment (kaladay) [reply src/main/resources/application.yaml:166]: I see I misread that.
`folio` is the value and not the environment name.
`ENV` is the environment variable!?
I think that is extremely bad practice and we should not reproduce bad practice.

This should be something other than the word `ENV`.
There is too much namespace conflict potential.

I now strongly ask for `FOLIO_ENV` or `ENV_FOLIO`.

Comment (wwelling) [reply src/main/resources/application.yaml:166]: No argument here. But would prefer it be a follow issue to fully document and test.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java (hunk 4 above)
Comment (wwelling) [reply src/main/java/org/folio/rest/camunda/controller/WorkflowController.java:31]: There was no logging provided for easily verifying normal operation. Previously with debug might be incorrect as hiding critical information about an active workflow that can indicate a scheduled workflow will be running. Perhaps we can have an issue for defining logging.

File: descriptors/ModuleDescriptor-template.json (hunk 2 above)
Comment (wwelling) [reply descriptors/ModuleDescriptor-template.json:6]: Sure. And I will include that issue in the staging to main PR for this sprint effort.

File: src/main/java/org/folio/rest/camunda/controller/WorkflowController.java (hunk 4 above)
Comment (wwelling) [reply src/main/java/org/folio/rest/camunda/controller/WorkflowController.java]: Updated.

**Pull request merged**